File: paas-metasvr/src/main/java/com/zzstack/paas/underlying/metasvr/autodeploy/util/VoltDBDeployUtils.java
package com.zzstack.paas.underlying.metasvr.autodeploy.util;

import com.zzstack.paas.underlying.metasvr.bean.PaasDeployFile;
import com.zzstack.paas.underlying.metasvr.bean.PaasInstance;
import com.zzstack.paas.underlying.metasvr.bean.PaasSsh;
import com.zzstack.paas.underlying.metasvr.consts.FixDefs;
import com.zzstack.paas.underlying.metasvr.dataservice.dao.MetaDataDao;
import com.zzstack.paas.underlying.metasvr.global.DeployLog;
import com.zzstack.paas.underlying.metasvr.metadata.CmptMeta;
import com.zzstack.paas.underlying.metasvr.singleton.MetaSvrGlobalRes;
import com.zzstack.paas.underlying.utils.FixHeader;
import com.zzstack.paas.underlying.utils.bean.ResultBean;
import com.zzstack.paas.underlying.utils.consts.CONSTS;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public class VoltDBDeployUtils {
    
    public static boolean deployVoltDBServer(JsonObject voltdb, String version, String hosts, String userName, String userPasswd,
            String logKey, String magicKey, ResultBean result) {
        
        String instId = voltdb.getString(FixHeader.HEADER_INST_ID);
        String sshId = voltdb.getString(FixHeader.HEADER_SSH_ID);
        String clientPort = voltdb.getString(FixHeader.HEADER_VOLT_CLIENT_PORT);
        String adminPort = voltdb.getString(FixHeader.HEADER_VOLT_ADMIN_PORT);
        String webPort = voltdb.getString(FixHeader.HEADER_VOLT_WEB_PORT);
        String internalPort = voltdb.getString(FixHeader.HEADER_VOLT_INTERNAL_PORT);
        String replicaPort = voltdb.getString(FixHeader.HEADER_VOLT_REPLI_PORT);
        String zkPort = voltdb.getString(FixHeader.HEADER_VOLT_ZK_PORT);
        String sitesPerHost = voltdb.getString(FixHeader.HEADER_SITES_PER_HOST);
        String kfactor = voltdb.getString(FixHeader.HEADER_KFACTOR);
        String memLimit = voltdb.getString(FixHeader.HEADER_MEM_LIMIT);
        String heartBeatTimeout = voltdb.getString(FixHeader.HEADER_HEARTBEAT_TIMEOUT);
        String temptablesMaxSize = voltdb.getString(FixHeader.HEADER_TEMPTABLES_MAXSIZE);
        String elasticDuration = voltdb.getString(FixHeader.HEADER_ELASTIC_DURATION);
        String elasticThroughput = voltdb.getString(FixHeader.HEADER_ELASTIC_THROUGHPUT);
        String queryTimeout = voltdb.getString(FixHeader.HEADER_QUERY_TIMEOUT);
        String procedureLoginfo = voltdb.getString(FixHeader.HEADER_PROCEDURE_LOGINFO);
        String memAlert = voltdb.getString(FixHeader.HEADER_MEM_ALERT);
        
        PaasInstance inst = MetaSvrGlobalRes.get().getCmptMeta().getInstance(instId);
        if (DeployUtils.isInstanceDeployed(inst, logKey, result)) return true;

        PaasSsh ssh = DeployUtils.getSshById(sshId, logKey, result);
        if (ssh == null) return false;
        String servIp = ssh.getServerIp();
        String sshName = ssh.getSshName();
        String sshPwd = ssh.getSshPwd();
        int sshPort = ssh.getSshPort();
        
        {
            String info = String.format("deploy voltdb: %s:%s, instId:%s", servIp, internalPort, instId);
            DeployLog.pubLog(logKey, info);
        }
        
        SSHExecutor ssh2 = new SSHExecutor(sshName, sshPwd, servIp, sshPort);
        if (!DeployUtils.initSsh2(ssh2, logKey, result)) return false;
        
        if (DeployUtils.checkPortUpPredeploy(ssh2, "voltdb", instId, servIp, clientPort, logKey, result)) {
            ssh2.close();
            result.setRetCode(CONSTS.REVOKE_NOK);
            result.setRetInfo("voltdb.client_port is in use");
            return false;
        }
        
        if (DeployUtils.checkPortUpPredeploy(ssh2, "voltdb", instId, servIp, adminPort, logKey, result)) {
            ssh2.close();
            result.setRetCode(CONSTS.REVOKE_NOK);
            result.setRetInfo("voltdb.admin_port is in use");
            return false;
        }
        
        if (DeployUtils.checkPortUpPredeploy(ssh2, "voltdb", instId, servIp, webPort, logKey, result)) {
            ssh2.close();
            result.setRetCode(CONSTS.REVOKE_NOK);
            result.setRetInfo("voltdb.web_port is in use");
            return false;
        }
        
        if (DeployUtils.checkPortUpPredeploy(ssh2, "voltdb", instId, servIp, internalPort, logKey, result)) {
            ssh2.close();
            result.setRetCode(CONSTS.REVOKE_NOK);
            result.setRetInfo("voltdb.internal_port is in use");
            return false;
        }
        
        if (DeployUtils.checkPortUpPredeploy(ssh2, "voltdb", instId, servIp, replicaPort, logKey, result)) {
            ssh2.close();
            result.setRetCode(CONSTS.REVOKE_NOK);
            result.setRetInfo("voltdb.repli_port is in use");
            return false;
        }
        
        if (DeployUtils.checkPortUpPredeploy(ssh2, "voltdb", instId, servIp, zkPort, logKey, result)) {
            ssh2.close();
            result.setRetCode(CONSTS.REVOKE_NOK);
            result.setRetInfo("voltdb.zk_port is in use");
            return false;
        }
        
        //获取文件,解压文件
        if (!DeployUtils.fetchAndExtractTGZDeployFile(ssh2, FixDefs.DB_VOLTDB_FILE_ID, FixDefs.DB_VOLTDB_ROOT, version, logKey, result))
            return false;
        
        //修改文件名
        PaasDeployFile deployFile = DeployUtils.getDeployFile(FixDefs.DB_VOLTDB_FILE_ID, logKey, result);
        String srcFileName = deployFile.getFileName();
        
        if (version == null || version.isEmpty()) {
            version = deployFile.getVersion();
        }
        if (srcFileName.indexOf(CONSTS.REG_VERSION) != -1 && version != null && !version.isEmpty()) {
            srcFileName = srcFileName.replaceFirst(CONSTS.REG_VERSION, version);
        }

        int i = srcFileName.indexOf(CONSTS.TAR_GZ_SURFIX);
        String oldName = srcFileName.substring(0, i);

        String newName = oldName + "_" + internalPort;
        if (!DeployUtils.rm(ssh2, newName, logKey, result)) {
            ssh2.close();
            return false;
        }
        if (!DeployUtils.mv(ssh2, newName, oldName, logKey, result)) {
            ssh2.close();
            return false;
        }
        if (!DeployUtils.cd(ssh2, newName, logKey, result)) {
            ssh2.close();
            return false;
        }
        
        // modify deployment.xml
        // <cluster sitesperhost="%SITES_PER_HOST%"
        //          kfactor="%KFACTOR%"/>
        // <heartbeat timeout="%HEARTBEAT_TIMEOUT%"/>
        // <user name="%ADMIN_NAME%" roles="dev,ops,administrator" password="%ADMIN_PWD%"/>
        // <user name="%USER_NAME%" roles="user" password="%USER_PASSWORD%"/>
        // <temptables maxsize="%TEMPTABLES_MAXSIZE%"/>
        // <elastic duration="%ELASTIC_DURATION%" throughput="%ELASTIC_THROUGHPUT%"/>
        // <query timeout="%QUERY_TIMEOUT%"/>
        // <procedure loginfo="%PROCEDURE_LOGINFO%"/>
        // <memorylimit size="%MEMORYLIMIT_SIZE%" alert="%MEMORYLIMIT_ALERT%"/>
        DeployLog.pubLog(logKey, "modify deployment.xml configures ......");
        String deploymentFile = "./deployment.xml";
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_SITES_PER_HOST, sitesPerHost, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_KFACTOR, kfactor, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_HEARTBEAT_TIMEOUT, heartBeatTimeout, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_ADMIN_NAME, FixDefs.VOLTDB_ADMIN_NAME, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_ADMIN_PWD, FixDefs.VOLTDB_ADMIN_PWD, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_USER_NAME, userName, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_USER_PASSWORD, userPasswd, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_TEMPTABLES_MAXSIZE, temptablesMaxSize, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_ELASTIC_DURATION, elasticDuration, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_ELASTIC_THROUGHPUT, elasticThroughput, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_QUERY_TIMEOUT, queryTimeout, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_PROCEDURE_LOGINFO, procedureLoginfo, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_MEMORYLIMIT_SIZE, memLimit, deploymentFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_MEMORYLIMIT_ALERT, memAlert, deploymentFile, logKey, result)) return false;
        
        // modify start.sh
        // nohup ./bin/voltdb start --dir=./database --host=%HOSTS% \
        //     --client=%VOLT_CLIENT_PORT% --admin=%VOLT_ADMIN_PORT% \
        //     --http=%VOLT_WEB_PORT% --internal=%VOLT_INTERNAL_PORT% \
        //     --replication=%VOLT_REPLI_PORT% --zookeeper=%VOLT_ZK_PORT% > /dev/null 2>&1 &
        String zkAddr = String.format("%s:%s", servIp, zkPort);
        String startFile = "./start.sh";
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_HOSTS, hosts, startFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_VOLT_CLIENT_PORT, clientPort, startFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_VOLT_ADMIN_PORT, adminPort, startFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_VOLT_WEB_PORT, webPort, startFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_VOLT_INTERNAL_PORT, internalPort, startFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_VOLT_REPLI_PORT, replicaPort, startFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_VOLT_ZK_PORT, zkAddr, startFile, logKey, result)) return false;

        // modify stop.sh
        // --host=%HOST% --user=%USER_NAME% --password=%USER_PASSWORD%
        String stopFile = "./stop.sh";
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_HOST, servIp, stopFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_USER_NAME, FixDefs.VOLTDB_ADMIN_NAME, stopFile, logKey, result)) return false;
        if (!DeployUtils.sedOnFailClose(ssh2, FixDefs.CONF_USER_PASSWORD, FixDefs.VOLTDB_ADMIN_PWD, stopFile, logKey, result)) return false;
        
        // init database
        String initDBCmd = "./bin/voltdb init --dir=./database --config=deployment.xml";
        if (!DeployUtils.execSimpleCmdOnFailClose(ssh2, initDBCmd, logKey, result)) return false;
        
        // start
        DeployLog.pubLog(logKey, "start voltdb-server ......");
        String cmd = String.format("./%s", FixDefs.START_SHELL);
        if (!DeployUtils.execSimpleCmdOnFailClose(ssh2, cmd, logKey, result)) return false;
        
        if (!DeployUtils.checkPortUpOnFailClose(ssh2, "voltdb", instId, servIp, internalPort, logKey, result)) return false;

        // update t_meta_instance is_deployed flag
        if (!MetaDataDao.updateInstanceDeployFlag(instId, FixDefs.STR_TRUE, result, magicKey)) {
            ssh2.close();
            return false;
        }
        
        ssh2.close();
        return true;
    }
    
    public static boolean undeployVoltDBServer(JsonObject voltdb, String version, String logKey, String magicKey, ResultBean result) {
        String instId = voltdb.getString(FixHeader.HEADER_INST_ID);
        String sshId = voltdb.getString(FixHeader.HEADER_SSH_ID);
        String internalPort = voltdb.getString(FixHeader.HEADER_VOLT_INTERNAL_PORT);
        
        PaasSsh ssh = DeployUtils.getSshById(sshId, logKey, result);
        if (ssh == null) return false;
        String servIp = ssh.getServerIp();
        String sshName = ssh.getSshName();
        String sshPwd = ssh.getSshPwd();
        int sshPort = ssh.getSshPort();
        SSHExecutor ssh2 = new SSHExecutor(sshName, sshPwd, servIp, sshPort);
        if (!DeployUtils.initSsh2(ssh2, logKey, result)) return false;
        {
            String info = String.format("start undeploy voltdb, inst_id:%s, serv_ip:%s, internalPort:%s", instId, servIp, internalPort);
            DeployLog.pubLog(logKey, info);
        }
        CmptMeta cmptMeta = MetaSvrGlobalRes.get().getCmptMeta();
        
        PaasInstance inst = cmptMeta.getInstance(instId);
        if (DeployUtils.isInstanceNotDeployed(inst, logKey, result)) return true;
        if (!DeployUtils.initSsh2(ssh2, logKey, result)) return false;
        
        PaasDeployFile deployFile = DeployUtils.getDeployFile(FixDefs.DB_VOLTDB_FILE_ID, logKey, result);
        String srcFileName = deployFile.getFileName();

        if (version == null || version.isEmpty()) {
            version = deployFile.getVersion();
        }
        if (srcFileName.indexOf(CONSTS.REG_VERSION) != -1 && version != null && !version.isEmpty()) {
            srcFileName = srcFileName.replaceFirst(CONSTS.REG_VERSION, version);
        }

        int i = srcFileName.indexOf(CONSTS.TAR_GZ_SURFIX);
        String oldName = srcFileName.substring(0, i);

        String newName = oldName + "_" + internalPort;
        String rootDir = String.format("%s/%s/%s", FixDefs.PAAS_ROOT, FixDefs.DB_VOLTDB_ROOT, newName);
        if (!DeployUtils.cdOnFailClose(ssh2, rootDir, logKey, result)) {
            return false;
        }
        
        // stop
        DeployLog.pubLog(logKey, "stop voltdb-server ......");
        String cmd = "./stop.sh";
        if (!DeployUtils.execSimpleCmdOnFailClose(ssh2, cmd, logKey, result)) {
            return false;
        }
        
        if (!DeployUtils.checkPortDownOnFailClose(ssh2, "voltdb", instId, servIp, internalPort, logKey, result)) {
            return false;
        }
        
        if (!DeployUtils.cdOnFailClose(ssh2, "..", logKey, result)) {
            return false;
        }
        if (!DeployUtils.rmOnFailClose(ssh2, newName, logKey, result)) {
            return false;
        }

        // update instance deploy flag
        if (!MetaDataDao.updateInstanceDeployFlag(instId, FixDefs.STR_FALSE, result, magicKey)) {
            ssh2.close();
            return false;
        }

        ssh2.close();
        return true;
    }
    
    public static String getVoltDBHosts(JsonArray voltdbServerArr, String logKey, ResultBean result) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < voltdbServerArr.size(); ++i) {
            JsonObject voltdb = voltdbServerArr.getJsonObject(i);
            String sshId = voltdb.getString(FixHeader.HEADER_SSH_ID);
            PaasSsh ssh = DeployUtils.getSshById(sshId, logKey, result);
            if (ssh == null)
                continue;
            
            String servIp = ssh.getServerIp();
            
            if (sb.length() > 0)
                sb.append(",");
            
            sb.append(servIp);
        }
        
        return sb.toString();
    }
    
    public static boolean createValidationTable(JsonObject voltdb, String version, String logKey, String magicKey, ResultBean result) {
        String instId = voltdb.getString(FixHeader.HEADER_INST_ID);
        String sshId = voltdb.getString(FixHeader.HEADER_SSH_ID);
        String internalPort = voltdb.getString(FixHeader.HEADER_VOLT_INTERNAL_PORT);
        String clientPort = voltdb.getString(FixHeader.HEADER_VOLT_CLIENT_PORT);
        
        PaasSsh ssh = DeployUtils.getSshById(sshId, logKey, result);
        if (ssh == null) return false;
        String servIp = ssh.getServerIp();
        String sshName = ssh.getSshName();
        String sshPwd = ssh.getSshPwd();
        int sshPort = ssh.getSshPort();
        SSHExecutor ssh2 = new SSHExecutor(sshName, sshPwd, servIp, sshPort);
        CmptMeta cmptMeta = MetaSvrGlobalRes.get().getCmptMeta();
        
        PaasInstance inst = cmptMeta.getInstance(instId);
        if (DeployUtils.isInstanceNotDeployed(inst, logKey, result)) return true;
        if (!DeployUtils.initSsh2(ssh2, logKey, result)) return false;
        
        PaasDeployFile deployFile = DeployUtils.getDeployFile(FixDefs.DB_VOLTDB_FILE_ID, logKey, result);
        String srcFileName = deployFile.getFileName();

        if (version == null || version.isEmpty()) {
            version = deployFile.getVersion();
        }
        if (srcFileName.indexOf(CONSTS.REG_VERSION) != -1 && version != null && !version.isEmpty()) {
            srcFileName = srcFileName.replaceFirst(CONSTS.REG_VERSION, version);
        }

        int i = srcFileName.indexOf(CONSTS.TAR_GZ_SURFIX);
        String oldName = srcFileName.substring(0, i);

        String newName = oldName + "_" + internalPort;
        String rootDir = String.format("%s/%s/%s", FixDefs.PAAS_ROOT, FixDefs.DB_VOLTDB_ROOT, newName);
        if (!DeployUtils.cdOnFailClose(ssh2, rootDir, logKey, result)) {
            return false;
        }
        
        // ./bin/sqlcmd --servers=127.0.0.1 --port=21212 --user=admin --password=a123456
        String sqlCmd = String.format("./bin/sqlcmd --servers=%s --port=%s --user=%s --password=%s ",
                servIp, clientPort, FixDefs.VOLTDB_ADMIN_NAME, FixDefs.VOLTDB_ADMIN_PWD);
        if (!DeployUtils.createValidationTable(ssh2, sqlCmd, logKey, result)) {
            return false;
        }
        
        ssh2.close();
        return true;
    }

}

File: paas-metasvr-go/pkg/deployutils/voltdb/voltdb_deploy_utils.go
package voltdb

import (
	"fmt"

	"github.com/maoge/paas-metasvr-go/pkg/consts"
	"github.com/maoge/paas-metasvr-go/pkg/dao/metadao"
	DeployUtils "github.com/maoge/paas-metasvr-go/pkg/deployutils"
	"github.com/maoge/paas-metasvr-go/pkg/global"
	"github.com/maoge/paas-metasvr-go/pkg/meta"
	"github.com/maoge/paas-metasvr-go/pkg/result"
)

func GetVoltDBHosts(voltdbServerArr []map[string]interface{}, logKey string,
	paasResult *result.ResultBean) string {

	result := ""
	for idx, voltdb := range voltdbServerArr {
		sshID := voltdb[consts.HEADER_SSH_ID].(string)
		ssh := meta.CMPT_META.GetSshById(sshID)
		if ssh == nil {
			continue
		}

		if idx > 0 {
			result += ","
		}
		result += ssh.SERVER_IP
	}

	return result
}

func CreateValidationTable(voltdb map[string]interface{}, version, logKey, magicKey string,
	paasResult *result.ResultBean) bool {

	return true
}

func DeployVoltDBServer(voltdb map[string]interface{}, version, hosts, userName, userPasswd, logKey, magicKey string,
	paasResult *result.ResultBean) bool {

	instId := voltdb[consts.HEADER_INST_ID].(string)
	sshId := voltdb[consts.HEADER_SSH_ID].(string)
	clientPort := voltdb[consts.HEADER_VOLT_CLIENT_PORT].(string)
	adminPort := voltdb[consts.HEADER_VOLT_ADMIN_PORT].(string)
	webPort := voltdb[consts.HEADER_VOLT_WEB_PORT].(string)
	internalPort := voltdb[consts.HEADER_VOLT_INTERNAL_PORT].(string)
	replicaPort := voltdb[consts.HEADER_VOLT_REPLI_PORT].(string)
	zkPort := voltdb[consts.HEADER_VOLT_ZK_PORT].(string)
	sitesPerHost := voltdb[consts.HEADER_SITES_PER_HOST].(string)
	kfactor := voltdb[consts.HEADER_KFACTOR].(string)
	memLimit := voltdb[consts.HEADER_MEM_LIMIT].(string)
	heartBeatTimeout := voltdb[consts.HEADER_HEARTBEAT_TIMEOUT].(string)
	temptablesMaxSize := voltdb[consts.HEADER_TEMPTABLES_MAXSIZE].(string)
	elasticDuration := voltdb[consts.HEADER_ELASTIC_DURATION].(string)
	elasticThroughput := voltdb[consts.HEADER_ELASTIC_THROUGHPUT].(string)
	queryTimeout := voltdb[consts.HEADER_QUERY_TIMEOUT].(string)
	procedureLoginfo := voltdb[consts.HEADER_PROCEDURE_LOGINFO].(string)
	memAlert := voltdb[consts.HEADER_MEM_ALERT].(string)

	ssh := meta.CMPT_META.GetSshById(sshId)
	if ssh == nil {
		paasResult.RET_CODE = consts.REVOKE_NOK
		paasResult.RET_INFO = consts.ERR_SSH_NOT_FOUND
		return false
	}

	inst := meta.CMPT_META.GetInstance(instId)
	if DeployUtils.IsInstanceDeployed(logKey, inst, paasResult) {
		return true
	}

	sshClient := DeployUtils.NewSSHClientBySSH(ssh)
	if !DeployUtils.ConnectSSH(sshClient, logKey, paasResult) {
		return false
	} else {
		defer sshClient.Close()
	}

	info := fmt.Sprintf("deploy voltdb: %s:%s, instId:%s", ssh.SERVER_IP, internalPort, instId)
	global.GLOBAL_RES.PubLog(logKey, info)

	checkPorts := []string{clientPort, adminPort, webPort, internalPort, replicaPort, zkPort}
	if DeployUtils.CheckPortsUpPredeploy(sshClient, checkPorts, logKey, paasResult) {
		return false
	}

	// DB_VOLTDB_FILE_ID -> 'voltdb-%VERSION%.tar.gz'
	if !DeployUtils.FetchAndExtractTgzDeployFile(sshClient, consts.DB_VOLTDB_FILE_ID, consts.DB_VOLTDB_ROOT, version, logKey, paasResult) {
		return false
	}

	oldName := DeployUtils.GetVersionedFileName(consts.DB_VOLTDB_FILE_ID, version, logKey, paasResult)
	newName := fmt.Sprintf("%s_%s", oldName, internalPort)

	if !DeployUtils.RM(sshClient, newName, logKey, paasResult) {
		return false
	}
	if !DeployUtils.MV(sshClient, newName, oldName, logKey, paasResult) {
		return false
	}
	if !DeployUtils.CD(sshClient, newName, logKey, paasResult) {
		return false
	}

	// modify deployment.xml
	// <cluster sitesperhost="%SITES_PER_HOST%"
	//          kfactor="%KFACTOR%"/>
	// <heartbeat timeout="%HEARTBEAT_TIMEOUT%"/>
	// <user name="%ADMIN_NAME%" roles="dev,ops,administrator" password="%ADMIN_PWD%"/>
	// <user name="%USER_NAME%" roles="user" password="%USER_PASSWORD%"/>
	// <temptables maxsize="%TEMPTABLES_MAXSIZE%"/>
	// <elastic duration="%ELASTIC_DURATION%" throughput="%ELASTIC_THROUGHPUT%"/>
	// <query timeout="%QUERY_TIMEOUT%"/>
	// <procedure loginfo="%PROCEDURE_LOGINFO%"/>
	// <memorylimit size="%MEMORYLIMIT_SIZE%" alert="%MEMORYLIMIT_ALERT%"/>
	global.GLOBAL_RES.PubLog(logKey, "modify deployment.xml configures ......")
	deploymentFile := "./deployment.xml"
	DeployUtils.SED(sshClient, consts.CONF_SITES_PER_HOST, sitesPerHost, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_KFACTOR, kfactor, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_HEARTBEAT_TIMEOUT, heartBeatTimeout, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_ADMIN_NAME, consts.VOLTDB_ADMIN_NAME, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_ADMIN_PWD, consts.VOLTDB_ADMIN_PWD, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_USER_NAME, userName, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_USER_PASSWORD, userPasswd, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_TEMPTABLES_MAXSIZE, temptablesMaxSize, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_ELASTIC_DURATION, elasticDuration, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_ELASTIC_THROUGHPUT, elasticThroughput, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_QUERY_TIMEOUT, queryTimeout, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_PROCEDURE_LOGINFO, procedureLoginfo, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_MEMORYLIMIT_SIZE, memLimit, deploymentFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_MEMORYLIMIT_ALERT, memAlert, deploymentFile, logKey, paasResult)

	// modify start.sh
	// nohup ./bin/voltdb start --dir=./database --host=%HOSTS% \
	//     --client=%VOLT_CLIENT_PORT% --admin=%VOLT_ADMIN_PORT% \
	//     --http=%VOLT_WEB_PORT% --internal=%VOLT_INTERNAL_PORT% \
	//     --replication=%VOLT_REPLI_PORT% --zookeeper=%VOLT_ZK_PORT% > /dev/null 2>&1 &
	zkAddr := fmt.Sprintf("%s:%s", ssh.SERVER_IP, zkPort)
	startFile := fmt.Sprintf("./%s", consts.START_SHELL)
	DeployUtils.SED(sshClient, consts.CONF_HOSTS, hosts, startFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_VOLT_CLIENT_PORT, clientPort, startFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_VOLT_ADMIN_PORT, adminPort, startFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_VOLT_WEB_PORT, webPort, startFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_VOLT_INTERNAL_PORT, internalPort, startFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_VOLT_REPLI_PORT, replicaPort, startFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_VOLT_ZK_PORT, zkAddr, startFile, logKey, paasResult)

	// modify stop.sh
	// --host=%HOST% --user=%USER_NAME% --password=%USER_PASSWORD%
	stopFile := fmt.Sprintf("./%s", consts.STOP_SHELL)
	DeployUtils.SED(sshClient, consts.CONF_HOST, ssh.SERVER_IP, stopFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_USER_NAME, consts.VOLTDB_ADMIN_NAME, stopFile, logKey, paasResult)
	DeployUtils.SED(sshClient, consts.CONF_USER_PASSWORD, consts.VOLTDB_ADMIN_PWD, stopFile, logKey, paasResult)

	// init database
	initDBCmd := "./bin/voltdb init --dir=./database --config=deployment.xml"
	if !DeployUtils.ExecSimpleCmd(sshClient, initDBCmd, logKey, paasResult) {
		return false
	}

	global.GLOBAL_RES.PubLog(logKey, "start voltdb-server ......")
	cmd := fmt.Sprintf("./%s", consts.START_SHELL)
	if !DeployUtils.ExecSimpleCmd(sshClient, cmd, logKey, paasResult) {
		return false
	}

	if !DeployUtils.CheckPortUp(sshClient, "dashboard-proxy", instId, internalPort, logKey, paasResult) {
		return false
	}

	// update instance deploy flag
	if !metadao.UpdateInstanceDeployFlag(instId, consts.STR_TRUE, logKey, magicKey, paasResult) {
		return false
	}

	return true
}

func UndeployVoltDBServer(voltdb map[string]interface{}, version, logKey, magicKey string,
	paasResult *result.ResultBean) bool {

	instId := voltdb[consts.HEADER_INST_ID].(string)
	sshId := voltdb[consts.HEADER_SSH_ID].(string)
	internalPort := voltdb[consts.HEADER_VOLT_INTERNAL_PORT].(string)

	inst := meta.CMPT_META.GetInstance(instId)
	if DeployUtils.IsInstanceNotDeployed(logKey, inst, paasResult) {
		return true
	}

	ssh := meta.CMPT_META.GetSshById(sshId)
	sshClient := DeployUtils.NewSSHClientBySSH(ssh)
	if !DeployUtils.ConnectSSH(sshClient, logKey, paasResult) {
		return false
	} else {
		defer sshClient.Close()
	}

	info := fmt.Sprintf("start undeploy voltdb-server, inst_id:%s, serv_ip:%s, http_port:%s", instId, ssh.SERVER_IP, internalPort)
	global.GLOBAL_RES.PubLog(logKey, info)

	oldName := DeployUtils.GetVersionedFileName(consts.DB_VOLTDB_FILE_ID, version, logKey, paasResult)
	newName := fmt.Sprintf("%s_%s", oldName, internalPort)
	rootDir := fmt.Sprintf("%s/%s/%s", consts.PAAS_ROOT, consts.DB_VOLTDB_ROOT, newName)
	if !DeployUtils.CD(sshClient, rootDir, logKey, paasResult) {
		return false
	}

	// stop
	global.GLOBAL_RES.PubLog(logKey, "stop voltdb-server ......")
	cmd := fmt.Sprintf("./%s", consts.STOP_SHELL)
	if !DeployUtils.ExecSimpleCmd(sshClient, cmd, logKey, paasResult) {
		return false
	}

	if !DeployUtils.CheckPortDown(sshClient, "voltdb-server", instId, internalPort, logKey, paasResult) {
		return false
	}

	DeployUtils.CD(sshClient, "..", logKey, paasResult)
	DeployUtils.RM(sshClient, newName, logKey, paasResult)

	// update instance deploy flag
	if !metadao.UpdateInstanceDeployFlag(instId, consts.STR_FALSE, logKey, magicKey, paasResult) {
		return false
	}

	return true
}
